blog/content/post/2021-10-28-rockchip-secure-boot.md

```diff
@@ -0,0 +1,364 @@
+for verifying Linux.
+
+To get Secure Boot working following things must be done.
+
+- Generate private and public keypair
+- Burn public key into eFUSE's
+- Sign `idbloader.img` (U-Boot TPL+SPL merged into one file)
+- Configure Verified Boot in SPL and U-Boot
+
+Let's begin.
+
+## Preparing
+
+RockChip provides proprietary binary-only tools for signing code and burning
+eFUSE's, code signing is handled by `rk_sign_tool` (Linux), Secure Boot Tool
+(Windows), eFUSE burning is done using eFUSE Tool. Since eFUSE Tool is Windows
+exclusive I had to setup Windows box side-by-side my Linux workstation.
+
+I checked out latest revisions of RockChip [tools](https://github.com/rockchip-linux/tools)
+and [rkbin](https://github.com/rockchip-linux/rkbin) repos.
+
+First, I had to generate keys for code signing.
+
+```shell
+tools/linux/rk_sign_tool/rk_sign_tool cc --chip 3288
+linux/rk_sign_tool/rk_sign_tool kk --out keys
+```
+
+## Signing code
+
+eFUSE Tool accepts only a signed binary as its input, so it must be extracting
+that public key from that binary, this is the same as `loader.bin` we pass to
+`rkdeveloptool`, from now on, I will call it loader.
+
+Loader can be quickly assembled using tools and config files from
+`rkbin/RKBOOT`. For example to build loader for RK3288, run following command
+(from `rkbin` directory)
+
+```shell
+tools/boot_merger RKBOOT/RK3288MINIALL.ini
+```
+
+I signed generated loader binary with `rk_sign_tool` without any problems.
+
+```shell
+$ ../tools/linux/rk_sign_tool/rk_sign_tool sl --key ../keys/privateKey.pem --pubkey ../keys/publicKey.pem --loader rk3288_loader_v1.09.258.bin
+start to sign rk3288_loader_v1.09.258.bin
+path = /hdd/rk-secure-boot/tools/linux/rk_sign_tool/temp/loader
+unpacking loader start...
+loader=rk3288_loader_v1.09.258.bin
+output=/hdd/rk-secure-boot/tools/linux/rk_sign_tool/temp/loader/
+unpacking UsbDataVRL...
+unpacking rk3288_ddr_400MHz_v1...
+unpacking rk3288_usbplug_v2...
+unpacking FlashData...
+unpacking FlashBoot...
+unpack loader ok.
+packing loader start...
+writing entry...
+writing /hdd/rk-secure-boot/tools/linux/rk_sign_tool/temp/loader/UsbDataVRL.bin
+writing /hdd/rk-secure-boot/tools/linux/rk_sign_tool/temp/loader/rk3288_ddr_400MHz_v1.bin
+writing /hdd/rk-secure-boot/tools/linux/rk_sign_tool/temp/loader/rk3288_usbplug_v2.bin
+writing /hdd/rk-secure-boot/tools/linux/rk_sign_tool/temp/loader/FlashData.bin
+writing /hdd/rk-secure-boot/tools/linux/rk_sign_tool/temp/loader/FlashBoot.bin
+writing crc...
+pack loader ok.(rk3288_loader_v1.09.258.bin)(0.01)
+sign loader ok.
+```
+
+But when I tried loading it into eFUSE Tool I've got an error.
+
+![Firmware loading failed](/img/rk_efuse_tool_error.png)
+
+That was an extremely helpful message. To find out what's wrong, I've done a
+quick analysis of unsigned and signed binary and discovered that the signed
+version has a different header - this is how the unsigned version looks like.
+
+```
```

Comment:
Maybe we can use some formatting here like C or other that can generate nice output?

Reply:
@pietrushnic You mean colored output? Rather not, I tried few options but it doesn't highlight properly all numbers (treats number with hex digits as normal words)
![Zrzut ekranu z 2021-11-10 16-59-56](https://user-images.githubusercontent.com/44615573/141147867-c4e55090-77b1-49d8-b697-4014c0ad3627.png)
Maybe we could roll our own solution?